Implement `FUNCTION` group of commands

elena-kolevska opened the pull request:
Implemented load, delete and flush. Working on `function list`

elena-kolevska commented:
Solves https://github.com/redis/go-redis/issues/2400, https://github.com/redis/go-redis/issues/2397 and https://github.com/redis/go-redis/issues/2395

monkey92t reviewed (commented):
Good, please continue...

elena-kolevska commented on command.go:
```diff
@@ -3840,3 +3840,167 @@ func (cmd *ZSliceWithKeyCmd) readReply(rd *proto.Reader) (err error) {
+			function.Description, err = rd.ReadString()
+			if err != nil {
+				if err == proto.Nil {
+					function.Description = "" // TODO Talk to Chayim
```
@monkey92t are we ok setting the function description to an empty string here? We're getting back a `redis: nil` but it didn't feel like we would need some kind of a nil object here, since it's just one property of a bigger object.

vmihailenco replied:
Can't say for @monkey92t , but it looks fine too me :+1:

elena-kolevska commented:
@vmihailenco , @monkey92t, if the approach I had with `FUNCTION LIST` is ok with you, I'll continue implementing the `WITHCODE` variant of `FUNCTION LIST` and all the rest of the `FUNCTIONS` group.

elena-kolevska commented on commands_test.go:
```diff
@@ -6086,6 +6086,159 @@ var _ = Describe("Commands", func() {
+			Expect(functionFlush.Err()).NotTo(HaveOccurred())
+
+			args := []interface{}{"function", "flush", "async"}
+			Expect(functionFlush.Args()).To(Equal(args))
```
If someone has a better idea of testing specifically the `async` flushing, I'd love to hear it. Otherwise, I think that just testing the command string and the end result (no functions on the server) should be good enough.

SoulPancake replied:
@elena-kolevska To specifically test the asynchronous behavior we can use a goroutine to perform some blocking operation, wait for a few seconds, and then call FunctionFlushAsync. We can then verify that the FunctionFlushAsync method does not block and returns without an error. Is that the right way to think?

elena-kolevska replied:
Yeah, great idea. Do you think we should add that to the test site, or would that fall more into the "testing the server" category?
I'm kind of inclined to think that for the commands that return a simple `OK` response where we don't need to populate a more complex object, checking that we've sent the correct string and we've gotten an OK should be good enough. What do you say?

elena-kolevska commented:
> `redis.RdsFunction` looks weird 😁

Indeed. I was reluctant to name it just `Function` though, cause even though it's not a reserved word in Go, when you're reading it in code, Redis Functions would probably not be the first thing to come to mind.
If you think that's not a concern though, I'm totally open to changing it.

monkey92t commented:
> > `redis.RdsFunction` looks weird 😁
> 
> Indeed. I was reluctant to name it just `Function` though, cause even though it's not a reserved word in Go, when you're reading it in code, Redis Functions would probably not be the first thing to come to mind. If you think that's not a concern though, I'm totally open to changing it.

`redis.Function{}` looks fine since go-redis only works with `redis` right? `redis.RdsFunction` looks weird as `pool.PoolName`.

elena-kolevska commented:
Ok, fixed. Thanks for the insight.

vmihailenco commented on commands.go:
```diff
@@ -395,6 +395,13 @@ type Cmdable interface {
+	FunctionDelete(ctx context.Context, libName string) *StringCmd
+	FunctionFlush(ctx context.Context) *StringCmd
+	FunctionFlushAsync(ctx context.Context) *StringCmd
+	FunctionList(ctx context.Context) *FunctionsListCmd
```
Let's rename FunctionsListCmd to FunctionListCmd for consistency with Redis `FunctionList`?

vmihailenco commented on command.go:
```diff
@@ -3840,3 +3840,167 @@ func (cmd *ZSliceWithKeyCmd) readReply(rd *proto.Reader) (err error) {
+func readRdsLibrarySlice(rd *proto.Reader) ([]Library, error) {
+	n, err := rd.ReadArrayLen()
+	if err != nil {
+		return []Library{}, err
```
nitpick: you can replace `[]Library{}` with `nil`

vmihailenco commented on command.go:
```diff
@@ -3840,3 +3840,167 @@ func (cmd *ZSliceWithKeyCmd) readReply(rd *proto.Reader) (err error) {
+		switch k {
+		case "library_name":
+			library.Name, err = rd.ReadString()
+			if err != nil {
```
looks like you double check the `err` here

vmihailenco commented:
Looks good so far :+1:

monkey92t commented:
I think:

```go
func FunctionList(libraryNamePattern string, withCode bool) []Library
```

We should try to maintain consistency with redis commands, and should not increase the cost for users to learn go-redis.

elena-kolevska commented:
Yeah, I definitely see the point there; I just feel that it wouldn't look very pretty when we call the method without the optional arguments:

```
redis.FunctionList("", false)
```

The following would actually be equivalent, and it looks better, but then we would have to teach people to use it that way; it's not obvious:
```
redis.FunctionList("*", false)
```

At the very minimum, I could make the library name pattern its own struct and pass nil. Still not great, but I think it looks a bit better than an empty string. What do you think?
```
redis.FunctionList(nil, false)

// Or with an argument:
redis.FunctionList(LibraryNamePattern{"mylib"}, false)
```

vmihailenco commented:
I think it should be either

```go
func FunctionList(libraryNamePattern string, withCode bool) []Library
```

or Option 5

```go
rdb.FunctionList(FunctionListArgs{
	LibraryName: "mylib",
	WithCode: true,
})				// []Library{}
```

The latter is somewhat more readable.

>I don't like returning a single Library object wrapped in an array when q.LibraryName is not null

We could add a helper function to retrieve a single library:

```go
rdb.FunctionList(...).Result() // []Library
rdb.FunctionList(...).First() // *Library
```

vmihailenco commented:
Also check:

```go
// SORT list LIMIT 0 2 ASC
vals, err := rdb.Sort(ctx, "list", &redis.Sort{Offset: 0, Count: 2, Order: "ASC"}).Result()

// ZRANGEBYSCORE zset -inf +inf WITHSCORES LIMIT 0 2
vals, err := rdb.ZRangeByScoreWithScores(ctx, "zset", &redis.ZRangeBy{
    Min: "-inf",
    Max: "+inf",
    Offset: 0,
    Count: 2,
}).Result()

// ZINTERSTORE out 2 zset1 zset2 WEIGHTS 2 3 AGGREGATE SUM
vals, err := rdb.ZInterStore(ctx, "out", &redis.ZStore{
    Keys: []string{"zset1", "zset2"},
    Weights: []int64{2, 3}
}).Result()
```

monkey92t commented:
I've considered "Option 5" too, if you all like it then we'll use it. 😄

elena-kolevska commented:
Great that we have an agreement :)

I implemented everything as you suggested, with the only difference being that the `.First()` method uses value instead of pointer semantics. I'm relatively new to Go, and I'm totally open to opinions here; I was just following the guideline that if it's a struct type and you're not expecting it to be mutated, you should use value semantics so you're not polluting the heap. As I said, though: new to Go, totally open to learning! ;)

I also implemented a `FirstVal()` method for completeness.

monkey92t commented:
Library is large enough that we can use ptr. What is the point of adding FirstVal()?

elena-kolevska commented:
Ok, I'll change it to a pointer then.

Re: FirstVal(), my reasoning was to have the matching "single-library" method for `Val()` as we have for `Result()`. I'm not a fan of the naming, though, I gotta admit.

```
rdb.FunctionList(...).Result() // []Library, err
rdb.FunctionList(...).First() // *Library, err

rdb.FunctionList(...).Val() // []Library
rdb.FunctionList(...).FirstVal() // *Library
```

monkey92t commented:
@elena-kolevska I tidied up the code.....I hope you won't mind. 😊

elena-kolevska commented:
Of course not! Quite the contrary, it’s great for my learning. 🙏

monkey92t commented:
@elena-kolevska  Do you want to add more commands? such: `function kill`, `function status`, or do we add it later?

elena-kolevska commented:
Since I'll only be able to come back to this next week, I'd suggest we close this PR now, and I'll add what's missing next week. There was also a PR on `FUNCTION KILL` already; we should check if we can use it.

monkey92t reviewed (approved):
OK, we don't need to close it, this PR can already do some work, we will add `function kill`, `function status` later.

monkey92t commented:
@elena-kolevska Waiting for you to finish the rest of the work. 😄

elena-kolevska commented:
I'm very much looking forward to it too ;)
Thank you all for your help with this PR.

vmihailenco commented:
@elena-kolevska thanks for your contribution and putting an extra effort into API design! It's clean and simple :+1:
